### PR TITLE
configurable data directory

### PR DESCRIPTION
**Pull request by rookgaard**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

Have you ever had thought you wanted to name each datapack for each server you have, while having the same engine? Now, using idea from 0.X, you can!

Changes I proposed replaces all occurences of hardcoded "data/" to string loaded from config file with default value as "data/", so from perspective of regular TFS user nothing will change, but give opportunity for those who (like me), wants to rename "data" directory.

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

**Comment by rookgaard** [comment]
@EPuncker Can you add reviewers please?

**Comment by MillhioreBT** [comment]
I think that with Git we can already handle different branches of the same project and just changing branches is enough to test different data packages, at least that's how I do it

**Comment by rookgaard** [comment]
> I think that with Git we can already handle different branches of the same project and just changing branches is enough to test different data packages, at least that's how I do it

That means, changes I proposed are exactly for you! :D

**Comment by EPuncker** [comment]
I like this idea because right now for example I have two different servers that use the same source but different datapacks and it will fit just nicely

**Review by MillhioreBT — APPROVED** [review]
Just because it is a more comfortable way to do it for anyone, I accept it!

File: src/baseevents.cpp [hunk 1]
```diff
@@ -34,7 +36,7 @@ bool BaseEvents::loadFromXml()
 	}
 
 	std::string scriptsName = getScriptBaseName();
-	std::string basePath = "data/" + scriptsName + "/";
+	std::string basePath = g_config.getString(ConfigManager::DATA_DIRECTORY) + scriptsName + "/";
```
Comment (ranisalt) [inline src/baseevents.cpp:39]: Please use a [`std::filesystem::path`](https://en.cppreference.com/w/cpp/filesystem/path) instead

File: src/chat.cpp [hunk 2]
```diff
@@ -287,10 +289,12 @@ Chat::Chat():
 
 bool Chat::load()
 {
+	std::string dataDirectory = g_config.getString(ConfigManager::DATA_DIRECTORY);
+	std::string chatchannelsFile = dataDirectory + "chatchannels/chatchannels.xml";
```
Comment (ranisalt) [inline src/chat.cpp:293]: Same as above

File: src/script.cpp [hunk 3]
```diff
@@ -41,7 +41,7 @@ bool Scripts::loadScripts(std::string folderName, bool isLib, bool reload)
 {
 	namespace fs = boost::filesystem;
```
Comment (ranisalt) [inline src/script.cpp:42]: You can now change this to be `std::filesystem`, we used `boost` because C++11 didn't have `filesystem` but C++17 has.

**Review by ranisalt — CHANGES_REQUESTED** [review]
The suggestion applies to all changes you made.

However, I suggest to pass the data folder in the command line (e.g. `tfs --data path/to/data` defaulting to `$CWD/data`)

**Comment by kornholi** [comment]
What do you think about a `ConfigManager::getDataPath` that returns `std::filesystem::path`? Then you could use `getDataPath() / "foo.xml"` etc.

**Comment by DSpeichert** [comment]
We have `data/`, `config.lua` and `key.pem`.

You'd still need to change the working directory to be able to choose different config.lua and (rarely) different key.pem.
So having "same engine", how is this better than specifying e.g. `WorkingDirectory=/path/to...` in a systemd unit?

**Comment by Zbizu** [comment]
Why not go further and create a login server?

**Comment by DSpeichert** [comment]
This seems abandoned after a number of thoughtful reviews, so I'm closing this. Feel free to respond or reopen.

**Pull request closed without merging**